### PR TITLE
bpo-31177: Skip deleted attributes while calling reset_mock

### PR DESCRIPTION
**Pull request by tirkarthi**
Skip the deleted attributes while calling `reset_mock` so that it doesn't cause an `AttributeError` .

Thanks

https://bugs.python.org/issue31177

### CODE REVIEW

File: Lib/unittest/test/testmock/testmock.py [hunk 1]
```diff
@@ -1566,6 +1566,22 @@ def test_attribute_deletion(self):
             self.assertRaises(AttributeError, getattr, mock, 'f')
 
 
+    def test_attribute_deletion_reset_mock(self):
+        mock = Mock()
```
Comment (mariocj89) [inline Lib/unittest/test/testmock/testmock.py:1572]: I would just leave this test as:

```
mock = Mock()
mock.child = True
del mock.child
mock.reset_mock()
```

And rename it to "test_reset_mock_does_not_raise_on_attr_deletion". The rest is verified in other test cases.

File: Misc/NEWS.d/next/Library/2018-09-14-10-38-18.bpo-31177.Sv91TN.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Skip deleted attributes while calling :meth:`mock.reset_mock`.
```
Comment (mariocj89) [inline Misc/NEWS.d/next/Library/2018-09-14-10-38-18.bpo-31177.Sv91TN.rst]: I'd suggest rewording this to something around:

"Fix bug that prevented using `reset_mock` on mock instances with deleted attributes"

**Review by mariocj89 — APPROVED** [review]
Great! Thanks a lot :)

@vstinner, this seems to fix a bug in unittest.mock. When attributes are deleted in a mock [they are marked with a mock.sentinel](https://github.com/python/cpython/blob/master/Lib/unittest/mock.py#L737) to prevent automatically creating other mocks dynamically, this is done by adding the _deleted sentinel to the child mocks dictionary.
The issue is that `_reset_mock` iterates over the list of child mocks assuming they are all Mocks without checking for the sentinel `_deleted`.

File: Lib/unittest/test/testmock/testmock.py [hunk 3]
```diff
@@ -1566,6 +1566,13 @@ def test_attribute_deletion(self):
             self.assertRaises(AttributeError, getattr, mock, 'f')
 
 
+    def test_reset_mock_does_not_raise_on_attr_deletion(self):
```
Comment (vstinner) [inline Lib/unittest/test/testmock/testmock.py:1569]: Would you mind to add a short comment inside the function explaining the test with a link to the bpo? ("# bpo-31177: (...)")

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by tirkarthi** [comment]
Please see https://github.com/python/cpython/pull/10807 that proposes deleted attributes to be retained that is different from my PR's behavior.

File: Lib/unittest/test/testmock/testmock.py [hunk 4]
```diff
@@ -1566,6 +1566,15 @@ def test_attribute_deletion(self):
...
+        # bpo-31177: reset_mock should not raise AttributeError when attributes
+        # were deleted in a mock instance
+        mock = Mock()
+        mock.child = True
+        del mock.child
+        mock.reset_mock()
```
Comment (vstinner) [inline Lib/unittest/test/testmock/testmock.py:1575]: Can you add a test to ensure that the attribute is still deleted after reset_mock()? For example, add:

```
self.assertFalse(hasattr(mock, 'child'))
```

**Comment by vstinner** [comment]
Oops, I had a draft comment that I forgot to send!

File: Lib/unittest/test/testmock/testmock.py (hunk 4 above)
Comment (tirkarthi) [reply Lib/unittest/test/testmock/testmock.py:1575]: Added an extra assertion. Thanks.

**Comment by cjw296** [comment]
https://github.com/python/cpython/pull/9302

**Comment by vstinner** [comment]
Thanks for the fix!
